dashboard/app/assets/javascripts/header.js

```diff
@@ -445,16 +445,28 @@ function mergedActivityCssClass(a, b) {
 }
 
 function populateProgress(scriptName) {
-  var scriptProgress = dashboard.clientState.allLevelsProgress()[scriptName] || {};
+  var clientProgress = dashboard.clientState.allLevelsProgress()[scriptName] || {};
+  $.ajax('/api/user_progress/' + scriptName).done(function (data) {
+    if (!data || !data.levels) {
+      return;
+    }
+    var serverProgress = data.levels;
 
-  // Aggregate progress from server and client
-  $.each(scriptProgress, function (level_id, result) {
-    var level_link = $('#level-' + level_id);
-    var status = mergedActivityCssClass(result, level_link.data('result'));
+    // Aggregate progress from server and client
+    var keys = Object.keys(clientProgress).concat(Object.keys(serverProgress));
+    var uniqueKeys = {};
+    keys.forEach(function (key) {
+      uniqueKeys[key] = true;
+    });
 
-    if (!level_link.hasClass(status)) {
-      // Clear the existing class and replace
-      level_link.attr('class', 'level_link ' + status);
-    }
+    Object.keys(uniqueKeys).forEach(function (levelId) {
+      var level_link = $('#level-' + levelId);
+      var status = mergedActivityCssClass(clientProgress[levelId], serverProgress[levelId].result);
```

Comment:
The code in question only sets an entry if the status on the server != 'not_tried'.  In that case serverProgress[levelId].result will result in a NPE because serverProgress[levelId] is null.

      if completion_status != 'not_tried'
        user_data[:levels][sl.level_id] = {
          status: completion_status,
          result: result
        }
      end